### PR TITLE
Initial sync machinery

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -19,13 +19,16 @@ include = [
 
 [tool.poetry.dependencies]
 python = "^3.11"
+attrs = "^23.1.0"
```
Comment (Schalk1e) [inline pyproject.toml:22]: Ah, noting attrs for future use.

File: src/aaq_sync/data_export_client.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def from_json(
+        cls, client: "ExportClient", table: str, resp_json: ResponseJSON
+    ) -> Self:
+        return cls(client, table, resp_json["metadata"], resp_json["result"])
+
+    def __iter__(self) -> TGen[JSONDict]:
+        yield from self.items
+
+    def iter_all(self) -> TGen[JSONDict]:
+        yield from self.items
```
Comment (Schalk1e) [inline src/aaq_sync/data_export_client.py:50]: Noting this as a better and more precise way to iterate.

File: src/aaq_sync/itertools.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+from collections.abc import Iterable, Iterator
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+
+class IteratorWithFinishedCheck(Generic[T]):
```
Comment (Schalk1e) [inline src/aaq_sync/itertools.py:7]: Noting its use in `sync` ❤️

File: tests/test_export_client.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+from aaq_sync.data_export_client import ExportClient
+from aaq_sync.data_models import FAQModel
+
+from .fake_data_export import FakeDataExport
+
+
+@pytest.fixture()
+def fake_data_export(httpx_mock):
+    return FakeDataExport(URL("https://127.0.0.100:1234/"), httpx_mock)
```
Comment (Schalk1e) [inline tests/test_export_client.py:15]: Noted usage of httpx_mock, would be great to use this in the rdw-ingestion-tools package. Adding ticket to my backlog.

**Review by Schalk1e — APPROVED** [review]
Not strictly qualified to review this, but instructive!

File: pyproject.toml (hunk 1 above)
Comment (jerith) [reply pyproject.toml:22]: I use `attrs` in pretty much everything. `dataclasses` in the stdlib is explicitly a simplified version of it. :-)

File: tests/test_export_client.py (hunk 4 above)
Comment (jerith) [reply tests/test_export_client.py:15]: I don't know if it's worth splitting the export client into a separate library (without the data model stuff) that both this and your ingestion tools can use, but if it isn't we can put a modified copy over there.

If we do the separate library thing, we should probably include `FakeDataExport` to make testing the things that use it easier.

**Pull request merged**